src/actix/certificate_helpers.rs

```diff
@@ -1,126 +1,168 @@
+use std::io;
+use std::sync::Arc;
 use std::time::{Duration, Instant};
 
-use openssl::error::ErrorStack;
-use openssl::ssl::{
-    SniError, SslAcceptor, SslAcceptorBuilder, SslContext, SslContextBuilder, SslFiletype,
-    SslMethod, SslVerifyMode,
-};
-use openssl::x509::store::X509StoreBuilder;
-use openssl::x509::X509;
 use parking_lot::RwLock;
+use rustls::server::{AllowAnyAuthenticatedClient, ClientHello, ResolvesServerCert};
+use rustls::sign::CertifiedKey;
+use rustls::{Certificate, RootCertStore, ServerConfig};
+use rustls_pemfile::Item;
 
+use super::with_buf_read;
 use crate::settings::{Settings, TlsConfig};
 
-struct SslContextHolder {
-    ssl_context: SslContext,
+/// A TTL based rotating server certificate resolver
+struct RotatingCertificateResolver {
+    /// TLS configuration used for loading/refreshing certified key
     tls_config: TlsConfig,
-    verify_client_cert: bool,
-    refresh_interval: Duration,
-    last_updated: Instant,
+
+    /// TTL for each rotation
+    ttl: Option<Duration>,
+
+    /// Current certified key
+    key: RwLock<CertifiedKeyWithAge>,
 }
 
-impl SslContextHolder {
-    fn new(
-        tls_config: &TlsConfig,
-        verify_client_cert: bool,
-        refresh_interval: Duration,
-    ) -> Result<SslContextHolder, ErrorStack> {
-        let ssl_context = build_ssl_context(tls_config, verify_client_cert)?;
-        Ok(SslContextHolder {
-            ssl_context,
-            tls_config: tls_config.clone(),
-            verify_client_cert,
-            refresh_interval,
-            last_updated: Instant::now(),
+impl RotatingCertificateResolver {
+    pub fn new(tls_config: TlsConfig, ttl: Option<Duration>) -> io::Result<Self> {
+        let certified_key = load_certified_key(&tls_config)?;
+
+        Ok(Self {
+            tls_config,
+            ttl,
+            key: RwLock::new(CertifiedKeyWithAge::from(certified_key)),
         })
     }
 
-    fn try_get_ssl_context(&self) -> Option<&SslContext> {
-        (self.last_updated.elapsed() < self.refresh_interval).then_some(&self.ssl_context)
+    /// Get certificate key or refresh
+    ///
+    /// The key is automatically refreshed when the TTL is reached.
+    /// If refreshing fails, an error is logged and the old key is persisted.
+    fn get_key_or_refresh(&self) -> Option<Arc<CertifiedKey>> {
+        // Return current key if TTL has not elapsed
+        let mut key = self.key.read().key.clone();
+        match self.ttl {
+            None => return Some(key),
+            Some(ttl) if !self.key.read().is_expired(ttl) => return Some(key),
+            Some(_) => {}
+        }
+
+        // Refresh key
+        match self.refresh() {
+            Ok(certified_key) => key = certified_key,
+            Err(err) => log::error!("Failed to refresh TLS certificate, keeping current: {err}"),
+        }
+        Some(key)
+    }
+
+    /// Refresh certificate key, update and return
+    fn refresh(&self) -> io::Result<Arc<CertifiedKey>> {
```

Comment:
We could reduce the overhead a bit by storing the `modified` `SystemTime` of the cert + key and check if it actually has changed to avoid re-reading the same file over and over again.

Reply:
This should only happen once per hour (TTL), so I think this change would be insignificant. The mtime is not 100% reliable anyway.